packages/nextjs/src/utils/handlers.ts

```diff
@@ -24,73 +30,114 @@ export const withSentry = (handler: NextApiHandler): WrappedNextApiHandler => {
-            if (req.query) {
-              // TODO get this from next if possible, to avoid accidentally replacing non-dynamic parts of the path if
-              // they match dynamic parts
-              for (const [key, value] of Object.entries(req.query)) {
-                reqPath = reqPath.replace(`${value}`, `[${key}]`);
-              }
+          const url = `${req.url}`;
+          // pull off query string, if any
+          let reqPath = stripUrlQueryAndFragment(url);
+          // Replace with placeholder
+          if (req.query) {
+            // TODO get this from next if possible, to avoid accidentally replacing non-dynamic parts of the path if
+            // they match dynamic parts
+            for (const [key, value] of Object.entries(req.query)) {
+              reqPath = reqPath.replace(`${value}`, `[${key}]`);
             }
-            const reqMethod = `${(req.method || 'GET').toUpperCase()} `;
-
-            const transaction = startTransaction(
-              {
-                name: `${reqMethod}${reqPath}`,
-                op: 'http.server',
-                ...traceparentData,
-              },
-              // extra context passed to the `tracesSampler`
-              { request: req },
-            );
-            currentScope.setSpan(transaction);
           }
+          const reqMethod = `${(req.method || 'GET').toUpperCase()} `;
+
+          const transaction = startTransaction(
+            {
+              name: `${reqMethod}${reqPath}`,
+              op: 'http.server',
+              ...traceparentData,
+            },
+            // extra context passed to the `tracesSampler`
+            { request: req },
+          );
+          currentScope.setSpan(transaction);
+
+          // save a link to the scope on the response, so that even if there's an error (landing us outside of
+          // the domain), we can still finish the transaction and attach the correct data to it
+          (res as ScopedResponse).__sentryScope = currentScope;
         }
+      }
 
+      try {
         return await handler(req, res); // Call original handler
       } catch (e) {
-        withScope(scope => {
-          scope.addEventProcessor(event => {
+        if (currentScope) {
+          currentScope.addEventProcessor(event => {
             addExceptionMechanism(event, {
               handled: false,
             });
-            return parseRequest(event, req);
+            return event;
           });
           captureException(e);
-        });
-        throw e;
-      } finally {
-        const transaction = getActiveTransaction();
-        if (transaction) {
-          transaction.setHttpStatus(res.statusCode);
-
-          transaction.finish();
-        }
-        try {
-          await flush(2000);
-        } catch (e) {
-          // no-empty
         }
+        throw e;
       }
     });
 
     return await boundHandler();
   };
 };
+
+type ResponseEndMethod = ScopedResponse['end'];
```

Comment:
I think two separate types might be overkill for it 😬

Reply:
Yeah, it may be.

I tend to do this when I'm writing wrapping functions, just so I can have this signature:

```
function wrapEndMethod(origEnd: ResponseEndMethod): WrappedResponseEndMethod {
```

because I feel like it just makes it so so clear what's happening, and because it means that when wrapped methods appear elsewhere, they're immediately identifiable as such by their type. Also, while these two may be the same _TypeScript_ type, they're nonetheless (to my mind) two different kinds (types? 😛) of things: one is a built-in node method, one's a monkeypatched thing we made up, and that distinction is one I find helpful.

(As an aside, I do wish TS would let you assign two aliases at once, so that instead of

```
type ResponseEndMethod = ScopedResponse['end'];
type WrappedResponseEndMethod = ResponseEndMethod
```

you could do

```
type WrappedResponseEndMethod = type ResponseEndMethod = ScopedResponse['end'];
// or, if necessary:
// type WrappedResponseEndMethod = (type ResponseEndMethod = ScopedResponse['end']);
```

because then I feel like having the two types wouldn't seem as noticeably "extra" somehow.)